// oae-transfer/lib/test/util.js
/*!
 * Copyright 2017 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');
var assert = require('assert');
var shortid = require('shortid');
var fs = require('fs');
var util = require('util');

// Util
var ContentTestUtil = require('oae-content/lib/test/util');
var DiscussionsTestUtil = require('oae-discussions/lib/test/util');

// API
var RestAPI = require('oae-rest');

// DAO
var Transfer = require('oae-transfer/lib/model').Transfer;


/* ================ CREATE TRANSFER ================ */

/**
 * Create a transfer
 *
 * @param  {RestContext}    restContext             The REST context to use for making requests
 * @param  {String}         emailOrigin             The email of the user who wants to create a transfer
 * @param  {String}         emailTarget             The emailTarget of the transfer to create
 * @param  {String}         idUserOrigin            The id of the user who wants to create a transfer
 * @param  {Function}       callback                Standard callback function
 * @param  {Folder}         callback.transfer...    All folders that were generated as separate callback parameters
 * @throws {AssertionError}                         Thrown if an error occurred generating the tranfer
 */
var assertCreateTransferSucceeds = module.exports.assertCreateTransferSucceeds = function(restContext, emailOrigin, emailTarget, idUserOrigin, callback) {
    RestAPI.Transfer.createTransfer(restContext, idUserOrigin, emailOrigin, emailTarget, function(err, createdTransfer) {
        assert.ok(createdTransfer);
        assert.equal(createdTransfer.emailOrigin, emailOrigin);
        assert.equal(createdTransfer.emailTarget, emailTarget);
        assert.equal(createdTransfer.idUserOrigin, idUserOrigin);
        return callback(createdTransfer);
    });
};

/* ================ GET TRANSFER ================ */

/**
 * Get a transfer, ensuring that it fails in a specified way
 *
 * @param  {RestContext}        restContext     The REST context to use when getting the transfer
 * @param  {String}             idUserOrigin    The id of the user who wants to make the transfer
 * @param  {Number}             httpCode        The expected failure HTTP code of the request
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request did not fail in the expected manner
 */
var assertGetTransferFails = module.exports.assertGetTransferFails = function(restContext, idUserOrigin, httpCode, callback) {
    RestAPI.Transfer.getTransferById(restContext, idUserOrigin, function(err, transfer) {
        assert.ok(err);
        assert.strictEqual(err.code, httpCode);
        assert.ok(!transfer);
        return callback();
    });
};

/**
 * Get a transfer, ensuring that the request is successful
 *
 * @param  {RestContext}        restContext     The REST context to use when getting the transfer
 * @param  {String}             idUserOrigin    The id of the user who wants to make the transfer
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request did not fail in the expected manner
 */
var assertGetTransferSucceeds = module.exports.assertGetTransferSucceeds = function(restContext, idUserOrigin, callback) {
    RestAPI.Transfer.getTransferById(restContext, idUserOrigin, function(err, transfer) {
        assert.ok(!err);
        assert.ok(transfer);
        assert.strictEqual(transfer.idUserOrigin, idUserOrigin);
        return callback(transfer);
    });
};

/* ================ DELETE TRANSFER ================ */

/**
 * Delete a transfer, ensuring that it fails in a specified way
 *
 * @param  {RestContext}        restContext     The REST context to use when deleting the transfer
 * @param  {String}             emailOrigin     The emailOrigin of the transfer to delete
 * @param  {String}             code            The transfer code of the transfer to delete
 * @param  {String}             idUserOrigin    The idUserOrigin of the transfer to delete
 * @param  {Number}             httpCode        The expected failure HTTP code of the request
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request did not fail in the expected manner
 */
var assertDeleteTransferFails = module.exports.assertDeleteTransferFails = function(restContext, emailOrigin, code, idUserOrigin, httpCode, callback) {
    RestAPI.Transfer.deleteTransfer(restContext, emailOrigin, code, idUserOrigin, function(err) {
        assert.ok(err);
        assert.strictEqual(err.code, httpCode);
        return callback();
    });
};

/**
 * Delete a transfer, ensuring that the request is successful
 *
 * @param  {RestContext}        restContext     The REST context to use when deleting the transfer
 * @param  {String}             emailOrigin     The emailOrigin of the transfer to delete
 * @param  {String}             code            The transfer code of the transfer to delete
 * @param  {String}             idUserOrigin    The idUserOrigin of the transfer to delete
 * @param  {Number}             httpCode        The expected failure HTTP code of the request
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request did not fail in the expected manner
 */
var assertDeleteTransferSucceeds = module.exports.assertDeleteTransferSucceeds = function(restContext, emailOrigin, code, idUserOrigin, callback) {
    RestAPI.Transfer.deleteTransfer(restContext, emailOrigin, code, idUserOrigin, function(err) {
        assert.ok(!err);
        return callback();
    });
};

/* ================ MAKE TRANSFER ================ */

/**
 * Execute transfer, ensuring that the request is successful
 *
 * @param  {RestContext}        restContext     The REST context to use when deleting the transfer
 * @param  {String}             emailOrigin     The emailOrigin of the transfer 
 * @param  {String}             code            The code write by the user
 * @param  {String}             emailTarget     The emailTarget of the transfer
 * @param  {String}             idUserTarget    The id of the user target
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var assertMakeTransferSucceeds = module.exports.assertMakeTransferSucceeds = function(restContext, emailOrigin, code, emailTarget, idUserTarget, callback) {
    RestAPI.Transfer.makeTransfer(restContext, emailOrigin, code, emailTarget, idUserTarget, function(err, managers) {
        assert.ok(!err);
        return callback(err, managers);
    });
};

/**
 * Execute transfer, ensuring that it fails in a specified way
 *
 * @param  {RestContext}        restContext     The REST context to use when deleting the transfer
 * @param  {String}             emailOrigin     The emailOrigin of the transfer 
 * @param  {String}             code            The code write by the user
 * @param  {String}             emailTarget     The emailTarget of the transfer
 * @param  {String}             idUserTarget    The id of the user target
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var assertMakeTransferFails = module.exports.assertMakeTransferFails = function(restContext, emailOrigin, code, emailTarget, idUserTarget, httpCode, callback) {
    RestAPI.Transfer.makeTransfer(restContext, emailOrigin, code, emailTarget, idUserTarget, function(err) {
        assert.ok(err);
        assert.strictEqual(err.code, httpCode);
        return callback();
    });
};

/* ================ UTIL ================ */

/**
 * Generate 3 discussions
 *
 * @param  {restCtx}            restCtx         The REST context to use when ti create the discussions 
 * @param  {String}             privacy         The privacy of the discussions
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateDiscussions = module.exports.generateDiscussions = function(restCtx, privacy, callback) {
    RestAPI.Discussions.createDiscussion(restCtx, 'name', 'description', privacy, null, null, function(err, discussion1) {
        assert.ok(!err);
        RestAPI.Discussions.createDiscussion(restCtx, 'name', 'description', privacy, null, null, function(err, discussion2) {
            assert.ok(!err);
            RestAPI.Discussions.createDiscussion(restCtx, 'name', 'description', privacy, null, null, function(err, discussion3) {
                assert.ok(!err);
                callback(null, discussion1, discussion2, discussion3);
            });
        });
    });
};

/**
 * Generate a discussion
 *
 * @param  {restCtx}            restCtx         The REST context to use when create the discussion
 * @param  {String}             privacy         The privacy of the discussion
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateDiscussion = module.exports.generateDiscussion = function(restCtx, privacy, callback) {
    RestAPI.Discussions.createDiscussion(restCtx, 'name', 'description', privacy, null, null, function(err, discussion) {
        assert.ok(!err);
        callback(null, discussion);
    });
};

/**
 * Generate 3 links
 *
 * @param  {restCtx}            restCtx         The REST context to use when create the links 
 * @param  {String}             privacy         The privacy of the links
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateLinks = module.exports.generateLinks = function(restCtx, privacy, callback) {
    RestAPI.Content.createLink(restCtx, 'name', 'description', privacy, 'google.com', null, null, null, function(err, link1) {
        assert.ok(!err);
        RestAPI.Content.createLink(restCtx, 'name', 'description', privacy, 'google.com', null, null, null, function(err, link2) {
            assert.ok(!err);
            RestAPI.Content.createLink(restCtx, 'name', 'description', privacy, 'google.com', null, null, null, function(err, link3) {
                assert.ok(!err);
                callback(null, link1, link2, link3);
            });
        });
    });
};

/**
 * Generate a link
 *
 * @param  {restCtx}            restCtx         The REST context to use when create the link 
 * @param  {String}             privacy         The privacy of the link
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateLink = module.exports.generateLink = function(restCtx, privacy, callback) {
    RestAPI.Content.createLink(restCtx, 'name', 'description', privacy, 'google.com', null, null, null, function(err, link) {
        assert.ok(!err);
        callback(null, link);
    });
};

/**
 * Generate 3 files
 *
 * @param  {restCtx}            restCtx         The REST context to use when create the files 
 * @param  {String}             privacy         The privacy of the files
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateFiles = module.exports.generateFiles = function(restCtx, privacy, callback) {
    RestAPI.Content.createFile(restCtx, 'name', 'description', privacy, getFileStream, null, null, function(err, file1) {
        assert.ok(!err);
        RestAPI.Content.createFile(restCtx, 'name', 'description', privacy, getFileStream, null, null, function(err, file2) {
            assert.ok(!err);
            RestAPI.Content.createFile(restCtx, 'name', 'description', privacy, getFileStream, null, null, function(err, file3) {
                assert.ok(!err);
                callback(null, file1, file2, file3);
            });
        });
    });
};

/**
 * Generate a file
 *
 * @param  {restCtx}            restCtx         The REST context to use when create the file
 * @param  {String}             privacy         The privacy of the file
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateFile = module.exports.generateFile = function(restCtx, privacy, callback) {
    RestAPI.Content.createFile(restCtx, 'name', 'description', privacy, getFileStream, null, null, function(err, file) {
        assert.ok(!err);
        callback(null, file);
    });
};

/**
 * Generate a file
 *
 * @param  {restCtx}            restCtx         The REST context to use when create the file
 * @param  {String}             privacy         The privacy of the file
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateFileOnFolder = module.exports.generateFileOnFolder = function(restCtx, privacy, idFolder, idGroup, callback) {
    RestAPI.Content.createFile(restCtx, 'name', 'description', privacy, getFileStream, [idGroup], null, [idFolder], function(err, file) {
        assert.ok(!err);
        callback(null, file);
    });
};


/**
 * Generate 3 collabdocs
 *
 * @param  {restCtx}            restCtx         The REST context to use when create the collabdoc
 * @param  {String}             privacy         The privacy of the collabdoc
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateCollabdocs = module.exports.generateCollabdocs = function(restCtx, privacy, callback) {
    RestAPI.Content.createCollabDoc(restCtx, 'name', 'description', privacy, [], [], [], [], function(err, collab1) {
        assert.ok(!err);
        RestAPI.Content.createCollabDoc(restCtx, 'name', 'description', privacy, [], [], [], [], function(err, collab2) {
            assert.ok(!err);
            RestAPI.Content.createCollabDoc(restCtx, 'name', 'description', privacy, [], [], [], [], function(err, collab3) {
                assert.ok(!err);
                callback(null, collab1, collab2, collab3);
            });
        });
    });
};

/**
 * Generate a collabdoc
 *
 * @param  {restCtx}            restCtx         The REST context to use when create the collabdoc
 * @param  {String}             privacy         The privacy of the collabdoc
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateCollabdoc = module.exports.generateCollabdoc = function(restCtx, privacy, callback) {
    RestAPI.Content.createCollabDoc(restCtx, 'name', 'description', privacy, [], [], [], [], function(err, collab) {
        assert.ok(!err);
        callback(null, collab);
    });
};

/**
 * Generate right for a content
 *
 * @param  {restCtx}            owner           The ower of the content
 * @param  {String}             contributor     The user
 * @param  {String}             right           The right to attribute to the user
 * @param  {String}             content         The content to update
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateRightContent = module.exports.generateRightContent = function(owner, contributor, right, content, callback) {
    var memberUpdates = {};
    memberUpdates[contributor.user.id] = right;
    ContentTestUtil.assertUpdateContentMembersSucceeds(owner.restContext, owner.restContext, content.id, memberUpdates, function(err) {
        assert.ok(!err);
        callback(null, content);
    });
};

/**
 * Generate right for a discussion
 *
 * @param  {restCtx}            owner           The ower of the discussion
 * @param  {String}             contributor     The user
 * @param  {String}             right           The right to attribute to the user
 * @param  {String}             discussion      The discussion to update
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateRightDiscussion = module.exports.generateRightDiscussion = function(owner, contributor, right, discussion, callback) {
    var memberUpdates = {};
    memberUpdates[contributor.user.id] = right;
    DiscussionsTestUtil.assertUpdateDiscussionMembersSucceeds(owner.restContext, owner.restContext, discussion.id, memberUpdates, function(err) {
        assert.ok(!err);
        callback(null, discussion);
    });
};

/**
 * Get the path
 *
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var getFileStream = module.exports.getFileStream = function() {
    var file = __dirname + '/data/profilepic.jpg';
    return fs.createReadStream(file);
};
